File: 2368-reachable-nodes-with-restrictions/2368-reachable-nodes-with-restrictions.py
class Solution:
    def reachableNodes(self, n: int, edges: List[List[int]], rst: List[int]) -> int:
        
        from collections import defaultdict 
        graph = defaultdict(list)
        seen = set(rst) 
        for x, y in edges:
            if (x in seen) or( y in seen):
                continue 
            graph[x].append(y)
            graph[y].append(x) 
        
        global ans 
        ans =0 
        def dfs(v):
            global ans 
            
            seen.add(v) 
            ans +=1 
            for nv in graph[v]: 
                if nv not in seen: 
                    dfs(nv) 



        
        dfs(0)
        return ans 

File: 2368-reachable-nodes-with-restrictions/2368-reachable-nodes-with-restrictions.cpp
class Solution {
public:
    int reachableNodes(int n, vector<vector<int>>& edges, vector<int>& restricted) {
        
        vector<bool> visit(n ,false) ; 
        vector<int> graph[n] ; 
        for(int i=0; i < edges.size() ; i++){
            graph[ edges[i][0] ].push_back(edges[i][1]);
            graph[ edges[i][1] ].push_back(edges[i][0]); 
        }
        for(auto r:restricted){
            visit[r] = true ; 
        }
        int ans = 0 ; 
        dfs(0 ,graph , ans , visit) ; 
        return ans ; 
        
    }
        
    void dfs(int v , vector<int> graph[] , int &ans , vector<bool> &visit){ 
        visit[v] = true ; 
        ans ++ ; 
        for(auto nv:graph[v]){
            if( visit[nv] == false){
                dfs(nv , graph , ans , visit) ;
            }
        }
}

        
};